homeworks/solutions/krasnor/4-param/src/halfedge.js

```diff
@@ -0,0 +1,588 @@
+          current.twin = boundaryHalfedge
+
+          current = nextHe
+        } while (current != he)
+        this.boundaries.push(boundaryCycle) // ## adding boundary cycle array
+
+        // link the cycle of boundary halfedges together
+        const n = boundaryCycle.length
+        for (let j = 0; j < n; j++) {
+          boundaryCycle[j].next = boundaryCycle[(j + n - 1) % n]
+          boundaryCycle[j].prev = boundaryCycle[(j + 1) % n]
+          hasTwinHalfedge.set(boundaryCycle[j], true)
+          hasTwinHalfedge.set(boundaryCycle[j].twin, true)
+        }
+      }
+    }
+    // allocate indices for all elements
+    let index = 0
+    this.vertices.forEach(v => {
+      v.idx = index++
+    })
+    index = 0
+    this.edges.forEach(e => {
+      e.idx = index++
+    })
+    index = 0
+    this.faces.forEach(f => {
+      f.idx = index++
+    })
+    index = 0
+    this.halfedges.forEach(he => {
+      he.idx = index++
+    })
+
+    console.log("finished mesh parsing")
+    console.log("boundary cycles: " + this.boundaries.length)
+    console.log("boundary[0] length: " + this.boundaries[0]?.length)
+    console.log("")
+  }
+
+  /**
+   * flatten computes the UV coordinates of the given triangle mesh.
+   *
+   * This implementation reuiqres the mesh contains at least one
+   * boundary loop.
+   *
+   * @param {string} boundaryType 'disk', or 'rect'
+   * @param {string} laplaceWeight 'uniform', or 'cotan'
+   */
+  flatten(boundaryType, laplaceWeight) {
+    // TODO: implement the Tutte's barycentric embedding algorithm.
+    //
+    // Hint:
+    // 1. check if the mesh contains a boundary
+    // 2. compute boundary uv coordinates depending on the boundary type
+    // 3. compute matrix depending on the laplacian weight type
+    // 4. solve linear equation and assing computed uv as vertex uv
+    //
+
+    // 1.
+    if(this.boundaries.length <= 0)
+      return;
+
+    // 2. compute boundary UVs
+    let U = DenseMatrix.zeros(this.vertices.length,1)
+    let V = DenseMatrix.zeros(this.vertices.length,1)
+
+    let n_bound_edges = this.boundaries[0].length
+    let uv_origin_vertex = -1;
+    switch (boundaryType) {
+      case 'disk':
+        this.boundaries[0].forEach((hf, i) => {
+          let u = (Math.cos(2*Math.PI * i/n_bound_edges))
+          let v = (Math.sin(2*Math.PI * i/n_bound_edges))
+          u = (u+1)/2 // normalize to [0,1]
+          v = (-v+1)/2
+          hf.vertex.uv.x = u
+          hf.vertex.uv.y = v
+          U.set(u, hf.vertex.idx,0)
+          V.set(v, hf.vertex.idx,0)
+        })
+        break
+      case 'rect':
+        const step_offset = 1/(n_bound_edges/4)
+        let corner_snap_stage = 0
+        let counter = 1
+
+        // console.log("rect case: hf: %f offset: %f",n_bound_edges,step_offset)
+        this.boundaries[0].forEach((hf, i) => {
+          let u = 0
+          let v = 0
+
+          // TODO rewrite
+          // works but not nice
```

Comment:
Interesting. I was thinking if you might come up with some nicer implementation (comparing to how I quickly wrote in the reference solution) for handling rectangle boundary vertex positions, but overall the current approach works fine :)